modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/IgniteStatisticsRepositoryImpl.java

```diff
@@ -0,0 +1,258 @@
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Statistics repository implementation.
+ */
+public class IgniteStatisticsRepositoryImpl implements IgniteStatisticsRepository {
+    /** Logger. */
+    private IgniteLogger log;
+
+    /** */
+    private final GridKernalContext ctx;
+
+    /** Table->Partition->Partition Statistics map, populated only on server nodes without persistence enabled.  */
+    private final Map<StatsKey, Map<Integer, ObjectPartitionStatisticsImpl>> partsStats;
+
+    /** Local (for current node) object statistics. */
+    private final Map<StatsKey, ObjectStatisticsImpl> localStats;
+
+    /** Global (for whole cluster) object statistics. */
+    private final Map<StatsKey, ObjectStatisticsImpl> globalStats = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     */
+    public IgniteStatisticsRepositoryImpl(GridKernalContext ctx) {
+        this.ctx = ctx;
+
+        if (ctx.config().isClientMode() || ctx.isDaemon()) {
+            // Cache only global statistics, no store
+            partsStats = null;
+            localStats = null;
+        } else {
+            if (GridCacheUtils.isPersistenceEnabled(ctx.config())) {
+                // Persistence store
+                partsStats = null;
+            } else {
+                // Cache partitions statistics, no store
+                partsStats = new ConcurrentHashMap<>();
+            }
+            localStats = new ConcurrentHashMap<>();
+        }
+        log = ctx.log(IgniteStatisticsRepositoryImpl.class);
+    }
+
+    @Override public void saveLocalPartitionsStatistics(StatsKey key,
+                                                        Collection<ObjectPartitionStatisticsImpl> statistics,
+                                                        boolean fullStat) {
+        if (partsStats != null) {
+            Map<Integer, ObjectPartitionStatisticsImpl> statisticsMap = new ConcurrentHashMap<>();
+            for (ObjectPartitionStatisticsImpl s : statistics) {
+                if (statisticsMap.put(s.partId(), s) != null)
+                    log.warning(String.format("Trying to save more than one %s.%s partition statistics for partition %d",
+                            key.schema(), key.obj(), s.partId()));
+            }
+
+            if (fullStat) {
+                partsStats.compute(key, (k, v) -> {
+                    if (v == null)
+                        v = statisticsMap;
+                    else
+                        v.putAll(statisticsMap);
+
+                    return v;
+                });
+            } else {
+                partsStats.compute(key, (k, v) -> {
+                    if (v != null) {
+                        for (Map.Entry<Integer, ObjectPartitionStatisticsImpl> partStat : v.entrySet()) {
+                            ObjectPartitionStatisticsImpl newStat = statisticsMap.get(partStat.getKey());
```

Comment:
1. "key" is known here and 'statisticsMap' is local object in the method,  so this line can be moved to outside the lamda.
2. seems, partStats.merge() method can be used instead.